include/kframework/lemmas/bitwise-simplification.k

```diff
@@ -20,85 +22,108 @@ module BITWISE-SIMPLIFICATION [symbolic]
+    rule maxUInt72 &Int N  => N requires #rangeUInt(72,  N) [simplification]
+    rule N &Int maxUInt64  => N requires #rangeUInt(64,  N) [simplification]
+    rule maxUInt64 &Int N  => N requires #rangeUInt(64,  N) [simplification]
+    rule N &Int maxUInt56  => N requires #rangeUInt(56,  N) [simplification]
+    rule maxUInt56 &Int N  => N requires #rangeUInt(56,  N) [simplification]
+    rule N &Int maxUInt48  => N requires #rangeUInt(48,  N) [simplification]
+    rule maxUInt48 &Int N  => N requires #rangeUInt(48,  N) [simplification]
+    rule N &Int maxUInt40  => N requires #rangeUInt(40,  N) [simplification]
+    rule maxUInt40 &Int N  => N requires #rangeUInt(40,  N) [simplification]
+    rule N &Int maxUInt32  => N requires #rangeUInt(32,  N) [simplification]
+    rule maxUInt32 &Int N  => N requires #rangeUInt(32,  N) [simplification]
+    rule N &Int maxUInt24  => N requires #rangeUInt(24,  N) [simplification]
+    rule maxUInt24 &Int N  => N requires #rangeUInt(24,  N) [simplification]
+    rule N &Int maxUInt16  => N requires #rangeUInt(16,  N) [simplification]
+    rule maxUInt16 &Int N  => N requires #rangeUInt(16,  N) [simplification]
+    rule N &Int maxUInt8   => N requires #rangeUInt(8,   N) [simplification]
+    rule maxUInt8 &Int N   => N requires #rangeUInt(8,   N) [simplification]
 
-    rule (X &Int Y) <Int Z      => true requires 0 <=Int X andBool 0 <=Int Y andBool (X <Int Z orBool Y <Int Z)          [simplification]
-    rule (X |Int Y) <Int pow256 => true requires 0 <=Int X andBool 0 <=Int Y andBool X <Int pow256 andBool Y <Int pow256 [simplification]
+    rule [bitwise-or-geq-zero] : 0 <=Int (X |Int Y) => true requires 0 <=Int X andBool 0 <=Int Y [simplification, smt-lemma]
+    rule [bitwise-and-geq-zero]: 0 <=Int (X &Int Y) => true requires 0 <=Int X andBool 0 <=Int Y [simplification, smt-lemma]
 
+    rule [bitwise-and-lt]:       (X &Int Y) <Int Z      => true requires 0 <=Int X andBool 0 <=Int Y andBool (X <Int Z orBool Y <Int Z)          [simplification]
+    rule [bitwise-or-lt-pow256]: (X |Int Y) <Int pow256 => true requires 0 <=Int X andBool 0 <=Int Y andBool X <Int pow256 andBool Y <Int pow256 [simplification]
+
+    // PETAR: We should probably add analogous simplifications for other
+    //        relevant powers of two, possibly generally.
     rule notMaxUInt160 &Int X => 0 requires #rangeUInt(160, X) [simplification]
     rule maxUInt160 &Int (X |Int (notMaxUInt160 &Int Y:Int)) => X requires #rangeUInt(160, X) andBool 0 <=Int Y [simplification]
 
+    rule [bitwise-and-asWord-tail]:
+      X &Int #asWord ( _ +Bytes B ) => X &Int #asWord ( B )
+      requires 0 <=Int X andBool X <Int 2 ^Int (8 *Int lengthBytes(B))
+      [simplification]
+
+    rule [bitwise-and-asWord-tail-shift]:
+      X &Int ( #asWord ( _ +Bytes B ) >>Int Z ) => X &Int ( #asWord ( B ) >>Int Z )
+      requires 0 <=Int X andBool X <Int 2 ^Int (8 *Int lengthBytes(B) -Int Z)
```

Comment:
```suggestion
      requires 0 <=Int X andBool X <Int 2 ^Int ((8 *Int lengthBytes(B)) -Int Z)
```

Clarity never hurts! Or maybe it does. But I prefer explicit parens.

Reply:
Yes, done.